Mainline kernel support for MMP3

lkundrak opened the pull request:
Hi,

I've made OpenFirmware and mainline Linux run on a MMP3-based machine I managed 
to get my hands on -- a Dell "Ariel" board. Perhaps not much more effort is needed
to make XO-4 run mainline Linux also. This pull request attempts to achieve that.

It starts with some fixes and then a fair amount of adjustments for DT as
understood by mainline kernel bindings, not dissimilar in spirit to what has
been done for XO-1.75.

Some commits that generalize things and a few new drivers are included,
to make the Ariel support possible.

Just prior to the actual Ariel support commit, an attempt to provide
compatibility with the OLPC OS is made. It's a bit of a kludge for the time
being, but might serve as a way to check how much effort would keeping
compatibility take. It is good enough to boot the OLPC OS with unmodified
3.5.7_xo4-20160412.0323.olpc.f9af9f7 kernel on my machine.

  commit 32cad21a7 ("olpc/4.0: optionally build DT for OLPC's 3.5 kernel")
  commit 7869b925c ("olpc: turn on olpc-compat? if alternate? is inspected on boot")


Apart from a review, at this point a hands-on test with an actual XO-4 would 
help. This image contains the today's linux-next build along with the binary
images of OpenFirmware and the OLPC OS kernel image:

  https://people.freedesktop.org/~lkundrak/mmp3.img.gz

The olpc.fth prints out details on what's there once booted from a USB flash
stick (or a SD card). If you could find time to give it a shot on actual XO-4
please share the dmesg and ideally the sysfs contents.

Note that there's a slight chance of ending up with a brick, but I've tested of 
XO-1.75 and MMP3-based Ariel, so I believe it to be reasonably safe. I guess 
you have a way to recover.

Thanks
Lubo

lkundrak committed:
cforth: update to a newer version

This switches to the github.com/MitchBradley repository and pins it to
the commit that is known to build well. It contains fixes for newer
compilers, native and cross-compilers.

The cforth image grew a couple of kilos, which is why move the shim 4k
further on XO-1.75.

Tested on XO-1.75 (MMP2) and Dell Ariel (MMP3).

lkundrak committed:
olpc/4.0: fix the mmp3 compatible string

"mrvl" prefix is incorrect and we don't have to use it here.

The 3.5-based OLPC kernel's machine support binds to "olpc,xo-cl4",
while mainline binds to "marvell,mmp3"

lkundrak committed:
mmp3: fix the cpus/ node

Include compatible strings for both CPU cores as well as the enable
method. Remove the MMP3 cruft from the x86/pc CPU node -- it's
irrelevant and we don't need to clutter it any more.

lkundrak committed:
olpc: split the mmp2 l2cache into a separate file

It actually is not correct for mmp3. Move it aside, then we can
conditionalize it.

lkundrak committed:
mmp3: add l2 cache controller

It's Tauros 3 on MMP3. Conditionally use it instead of Tauros 2 on that
platform.

lkundrak committed:
olpc/cl4: switch SDHCI to clock from /clocks

lkundrak committed:
mmp3/galcore: switch to clock from /clocks

lkundrak committed:
mmp2: drop /apbc and /pmua

Now that there are unused, we can drop them. All clocks are provided by
the /clocks node.

lkundrak committed:
mmp3: switch /thermal to clock from /clocks

lkundrak committed:
mmp2: make /interrupt-controller compatible with marvell,mmp3-intc

Described by interrupt-controller/mrvl,intc.txt bindings.

lkundrak committed:
mmp3: add /snoop-control-unit

This is required to bring SMP up. Described by arm/scu.txt binding.

lkundrak committed:
mmp3/usb2phy: make /usb2-phy compatible with marvell,mmp3-usb-phy

This allows the Linux PHY driver to bind and be used with the EHCI
controller driver. Described by the phy/phy-mmp3-usb.txt binding.

lkundrak committed:
spiflash: add Windbond W25Q32FV chip

A 4M SPI flash chip as used on the Dell Ariel board.

lkundrak committed:
mmp3/ariel: add Dell Ariel board support

This is the board used in Dell Wyse 3020 thin client. It's based on
MMP3, reusing most of OLPC XO-4 support.

lkundrak committed:
mmp2/twsi: move node creation from arm/olpc

Move twsi-i2c.fth out from arm/olpc/ into arm/mmp2/ and have it create
all the device nodes. If the platform doesn't need any, it can disable
them, just like olpc/build-fw.fth does.

Having the device around makes it possible for code like arm/mmp3/gic.fth to
just always assume the device nodes are always around.

lkundrak committed:
mmp2/clk: add clock IDs for TWSI3 and TWSI5

lkundrak committed:
mmp2/twsi-node: fix the register range

The TWSI register block doesn't take up the whole page. In fact, on
MMP3 the TWSI ranges are too close together so that they would overlap
if they were 4K long:

  pxa2xx-i2c d4033800.i2c: can't request region for resource
    [mem 0xd4033800-0xd40347ff]

lkundrak committed:
mmp2/clk: add SSP4 clock ID

lkundrak committed:
olpc/lcd: fix a handful of places that assume 16 bpp

lkundrak committed:
fdt: Fix the node name strings

Previously the tree would look like this:

  "/"
    "/subnode@0"
      "/subnode@0/leaf0"
    "/subnode@1"

Whereas the correct version would be:

  ""
    "subnode@0"
      "leaf0"
    "subnode@1"

lkundrak committed:
mmp2/uart: make the MMP UART compatible with intel,xscale-uart

The UART is XScale compatible and the more specific compatible string
shouldn't really have been necessary.

There are three drivers that bind to mrvl,mmp-uart in Linux, for
legacy reasons. The 8250_of driver should be preferrably used and the
other two sometimes break.

We need to keep the more specific compatible string for OLPC builds that
lack the 8250_of driver, but prefer the XScale one.

lkundrak committed:
Revert "fdt: Fix the node name strings"

phandle>basename sometimes causes infinite loops; no idea why. Drop that
abomination. We'll fix things differently.

This reverts commit a00c13115bf55c5caec907369584695c652fb03d.

lkundrak committed:
mmp2/clk: add a different compatible string on MMP3

There are hardware differencies, such as more clocks available on the newer
model. The driver needs to know which platform are we on.

lkundrak committed:
arm/olpc: fix gitignore

Don't ignore the build directory -- ignore the build artifacts
themselves as seems to have been intended.

lkundrak committed:
mmp3/gic: don't create separate interrupt mux nodes

The interrupt muxes we use are the same as when booted in non-GIC mode.
Just modify their "interrupts" property for the three-cell mode as well.

Otherwise Linux would get super confused.

lkundrak committed:
olpc/lcd: fix compatible string on MMP3

The hardware is slightly different (a different clock select register)
and the driver needs to know.

lkundrak committed:
mmp3/thermal: only show battery stats if there's a battery

lkundrak committed:
olpc/lcd: only call into DCON on OLPC

lkundrak committed:
olpc/usb: make usb-hub-reset-gpios property optional

Only create it if the board actually has a USB hub reset line.

lkundrak committed:
mmp2/rtc: conditionalize the suspend/resume testing machinery

Only include it if the OLPC EC is around, that can report the power
consumption.

lkundrak committed:
arm/olpc: move font selection from lcd.fth to build-fw.fth

The console font is not a function of the LCD controller.

lkundrak committed:
arm/olpc: split OLPC LCD configuration from the LCDC driver

Making the driver useful for boards configured differently than
OLPC.

lkundrak committed:
olpc/lcd: fix a messed up horizontal front porch calculation

There's a nice hfp word that gives us exactly the value we need.
Tested not to blow up on a XO-1.75.

lkundrak committed:
olpc: parametrize the gridmap and life demo resolution

On MMP2/MMP3 we have hdisp and vdisp words that describe the display size.
Otherwise still assume the OLPC resolution.

lkundrak committed:
olpc/twsi-i2c: fix TWSI5 address

In a comment, but still -- the documentation better be correct.

lkundrak committed:
mmp3/gic: delete the interrupt property before adding it anew

lkundrak committed:
mmp2: add nodes for all SSP SPI devices

Currently OLPC lumps together SSP1 and a SPI flash device, which
is impractical for boards that use more SSP blocks.

For now this doesn't implement the I/O driver; merely creates the device
nodes.

lkundrak committed:
mmp2/uart: factor out the UART nodes

lkundrak committed:
mmp3/gic: only fix up OLPC specific nodes on OLPC

lkundrak committed:
olpc/sdhci: split out MMP2 SDHCI code

This separates common MMP2/MMP3 SDHCI fixture from OLPC specifics. It
allows reuse from non-OLPC MMP boards.

lkundrak committed:
mmp3: add HSIC USB driver

lkundrak committed:
mmp3: tidy up interrupt controller references

Drop unnecessary interrupt-parent.

Qualify the names with the bus location so that it's obvious which
interrupt controller do we mean (there also are muxes and possibly a GIC).

Don't include INTC in MUX paths. It is there by accident -- the MUXes
cna happily multiplex interrupts via the GIC as well. This will make it
easier to tidy things up.

lkundrak committed:
mmp2/clk: fix the galcore clock numbers

Use the same numbers as Linux 5.7 uses.

lkundrak committed:
mmp2/galcore: drop assigned clocks

This was a bad idea altogether. The driver should decide which mux to
use or keep whatever firmware set. Moreover would require kernel to
expose the muxes by allocating a clock ids for them.

lkundrak committed:
olpc/usb: also initialize HSIC

Not on OLPC actually, since there HSIC is not used, but generally on
MMP3. (This file prehaps really belongs to cpu/mmp2, but hey...)

lkundrak committed:
mmp3/gic: create the TWD nodes

Arm TWD is used for local timer interrupt delivery on SMP systems. Enable it
and create the appropriate nodes in the GIC fixup path, because it relies on
GIC.

Apparently the non-GIC interrupt controller can be used for CPU local
timer interrupt delivery, but not with mainline Linux at this point.

lkundrak committed:
mmp2/clk: don't enable SD master clock along with SDHCI1

It's pretty arbitrary, only because the knob appears in the same
register. The master clock is actually also enabled in
cpu/arm/olpc/sdhci.fth:

  159 stand-init: SDHC clocks
  160    h# 400 h# 54 pmua!    \ Master SDH clock divisor

lkundrak committed:
usb2/net: add a driver for SMSC LAN7500

This is a driver for the USB 2.0 to 10/100/1000 Mbps Ethernet driver. At
the time it is pretty basic, it accepts all packets and makes no attempt
at filtering in hardware but works well enough.

lkundrak committed:
usb2/net: remove wrap-msg

It does nothing, is not used, and it's not clear how would it have
worked for devices that wrap ethernet data in their custom headers.

lkundrak committed:
usb2/net: add length-header callback

This allows the drivers to prepend extra headers to the packet data for
hardware that requires that. It is a more flexible replacement for
length-header? logic that only prepends Pegasus-style length header.

lkundrak committed:
usb2/net: drop extra copy-packet word

This one seems to be Pegasus specific and just forgotten in place. The one that
follows immediately shadows it.

lkundrak committed:
dev: add mv88pm867 driver

This is a regulator that is used on Dell Ariel (Wyse 3020) board to
configure DRAM voltage. For now the driver actually does nothing but create
the device node.

lkundrak committed:
video/dacs: add driver for Chrontel CH7033

This is a driver for video encoder with VGA and DVI/HDMI outputs.

There is no documentation for the chip -- the operation was guessed from
what was sniffed on a Dell Wyse 3020 ThinOS terminal with some clues from
largely obfuscated Mediatek's GPL code dump.

The configuration is hardcoded to what makes sense for 1024x768@60.
However, with VGA and clock bypasses on, it doesn't seem to matter too
much and other modes seem to work fine, which is good.

lkundrak committed:
dev: add driver for KB3930 Embedded Controller on Dell Ariel

This is an ENE Embedded Controller with the Dell Ariel firmware.
It seems to be able to control LEDs and USB port power.

lkundrak committed:
arm/linux: drop load-base off stack if not a zImage

At the time we check the zImage magic, there's ( adr len ) on the stack.
Pop both off if the magic check fails.

Otherwise the load of a different kind of image (such as Forth text larger
than 30 bytes -- we check the length on the previous line) would leave the
load base on stack which is just not nice.

Fixes: 2d7dad95a1f4 ("arm/linux: ignore the size from the zImage header")

lkundrak committed:
olpc/rtc: move into dev/ds1338

This driver is not particularly OLPC specific. Move it to dev/ and
unwrap it from /i2c@d4031000 to allow attaching it to different I2C
hosts.

lkundrak committed:
mmp3/thermal: add /thermal-zones node

This describes the thermal zones as specified by the thermal/thermal.txt
binding.

lkundrak committed:
mmp3/thermal: also pass the complete register value to .tc

Aside from the temperature bits, it also needs the lowrange bit to tell
which gray code table to look up the temperature value in. Otherwise it
always uses the high one; and then it appears that the CPU is on fire:

  ok .ts
  tsense_reg[0]=a4200d00
      31=reserved 29=status 26=en_wdog 21=auto_read_en
      wdog_tshld=d 100.5C
      int_tshld=0 78.0C
      temp_value=0 78.0C
  tsense_reg[1]=a8200009
      31=reserved 29=status 27=lowrange 21=auto_read_en
      wdog_tshld=0 78.0C
      int_tshld=0 78.0C
      temp_value=9 113.0C
  tsense_reg[2]=a8200009
      31=reserved 29=status 27=lowrange 21=auto_read_en
      wdog_tshld=0 78.0C
      int_tshld=0 78.0C
      temp_value=9 113.0C
  ok

lkundrak committed:
mmp3/thermal: add #thermal-sensor-cells

This is required as specified in the thermal/thermal.txt binding.

lkundrak committed:
mmp2/clk: add clock IDs for HSICs

lkundrak committed:
dev/usb2: use @1 for root hub port index

A hub is an USB device and devices' "port numbers range from 1 to 255"
according to Open Firmware, Recommended Practice: Universal Serial Bus (Version
1, June 1, 1998).

Linux thinks the same and if we kept indexing from zero it wouldn't
associate the Device Tree firmware node with the device instance.

Before:

  USB devices:
  /usb@d4208000/hub@0,0
  /usb@d4208000/hub@0,0/scsi@4,0
  /usb@d4208000/hub@0,0/ethernet@3,0
  /usb@d4208000/hub@0,0/scsi@4,0/disk

After:

  USB devices:
  /usb@d4208000/hub@1,0
  /usb@d4208000/hub@1,0/scsi@4,0
  /usb@d4208000/hub@1,0/ethernet@3,0
  /usb@d4208000/hub@1,0/scsi@4,0/disk

lkundrak committed:
mmp2/random: create a device node

This way the Linux driver can bind.

lkundrak committed:
thunderstone: switch vmeta to clock from /clocks

lkundrak committed:
thunderstone: switch /usb to clock from /clocks

lkundrak committed:
thunderstone: switch UART to clock from /clocks

lkundrak committed:
olpc/4.0: optionally build DT for OLPC's 3.5 kernel

Many of the preceding changes were done for the sake of compatibility
with mainline kernel and were done without considering compatibility
with OLPC XO-4.0's kernel, based on linux-3.5.

Most notably, the clock driver uses completely different DT bindings.
There are a couple more things -- some drivers are crashy when they
don't find the expected nodes, such as I2C node aliases or dumb panel.

This adds an "olpc-compat-" drop-in whose purpose is to beat the tree
back into shape that's acceptable for the non-mainline kernel.
It's loaded just prior to booting the kernel if olpc-compat? flag is
set.

lkundrak committed:
olpc/1.75/lcdcfg: keep the /panel node as it is

These were only meaningful for the OLPC kernel's pxa688fb driver, and
that one is going to be looking for them elsewhere -- in /display child
nodes. Drop them from here.

lkundrak committed:
olpc: turn on olpc-compat? if alternate? is inspected on boot

The OLPC OS on XO-4 (be it in secured mode, or on unsecured laptop booted
from an USB flash stick) executes the alternate? word to find out if it
shall boot the Android kernel or whatever it is.

While probably quite half-hearded, that seems to be the only way to tell if
we're running OLPC OS, that probably has a 3.5-based kernel that needs a
compatibility device tree.

The other distros that are potentially going to ship newer kernels have
no need to call alternate? in their boot paths.

lkundrak commented:
> It is good enough to boot the OLPC OS with unmodified 3.5.7_xo4-20160412.0323.olpc.f9af9f7 kernel on my machine.

Not kidding:

![olpcos](https://user-images.githubusercontent.com/92695/77363152-7a9bd400-6d52-11ea-9725-e5963ae035a7.jpeg)

lkundrak commented:
The binaries, test script and the device trees I've used for testing/verification: https://github.com/lkundrak/openfirmware/commit/0ae805907

quozl commented:
Tested flashing of mmp3.img.gz:xo4.rom without problems seen so far;

```log
ok dir u:\*.rom
ext2-file-system
----rw-r--r--   4194304  2020-03-23 20:27:45  ariel.rom
----rw-r--r--   1048576  2020-03-23 20:27:45  xo1.rom
----rw-r--r--   1048576  2020-03-23 20:27:45  xo175.rom
----rw-r--r--   2097152  2020-03-23 20:27:45  xo4.rom
ok flash u:\xo4.rom
Reading u:\xo4.rom
Got firmware version: CL4   Q7C07  Q7C
Checking integrity ...
SPI FLASH is type 35.3525c2, MX25U1635E, 2 MB.
Writing
1e0000
Writing

Verifying
1e0000
Verifying


CForth built 2020-03-23 18:38 from commit d3370770...
releasing
Undefined word encountered
USB devices:
/usb@d4208000/hub@1,0
OLPC 4B1, 1196 MHz, 1 GiB memory, 8 GB internal storage, S/N SHC23800036
OpenFirmware  Q7C07   EC Firmware 0.5.01   2020-03-24 04:05:32 UTC

Type the Esc key to interrupt automatic startup
2
OLPC 4B1, 1196 MHz, 1 GiB memory, 8 GB internal storage, S/N SHC23800036
OpenFirmware  Q7C07   EC Firmware 0.5.01   2020-03-24 04:05:32 UTC

Type 'help' for more information.

ok cr

```

How to verify?  I see version unchanged.

quozl commented:
Can't get at the USB drive any more.

```log
ok dir u:\
Can't open directory
```

quozl commented:
Reviewed your changes.  Let me know if there's something more I can do.

lkundrak commented:
Thank you.

> CForth built 2020-03-23 18:38 from commit d3370770...
releasing
> Undefined word encountered
> USB devices:
> /usb@d4208000/hub@1,0

I guess I'll follow up with an updated version when I figure out which word was undefined. Or, possibly, with some tracing, such as `stand-init-debug?` to narrow things down.

quozl commented:
Debugging session.  Here it is;

```log
ok dev /dcon
ok see stand-init
: stand-init 
   stand-init lose 0= if
      " /dcon" find-device 0 0 " no-freeze" 
      property device-end 
   then
;
ok see lose
: lose 
   true abort" Undefined word encountered " 
;
ok █
```

I've looked at mmp2dcon.fth in your branch, but can't figure out what happened.  I'm interrupted by other things just now.  Check build logs?

quozl commented:
I'm guessing `has-dcon-ram?` is now defined in a context unreachable by the code inside the `begin-package`.  It is not resolvable at the ok prompt, per `sifting`.

lkundrak committed:
olpc/1.75/lcfcfg: deselect root node when we're done with things

Otherwise words defined afterwards, such as has-dcon-ram? will end up
being root device's methods, resulting in no end of mayhem and suffering.

lkundrak commented:
> Check build logs?

Ah, should've done that before.

> I'm guessing has-dcon-ram? is now defined in a context unreachable by the code inside the begin-package

That is indeed correct. Should be fixed by:

  commit f484cbd1a ("olpc/1.75/lcfcfg: deselect root node when we're done with things")

The updated binary images are here: https://people.freedesktop.org/~lkundrak/ofw-mmp3-v2/

quozl commented:
Thanks.  Reflashed over wireless.  Undefined word no longer occurs.

Reflash of OLPC OS 13.2.11 using a USB Flash drive does work.  But it does not boot, serial console shows;

```log
CForth built 2020-03-24 09:27 from commit d3370770...
releasing
USB devices:
/usb@d4208000/hub@1,0
OLPC 4B1, 1196 MHz, 1 GiB memory, 8 GB internal storage, S/N SHC23800036
OpenFirmware  Q7C07   EC Firmware 0.5.01   2020-03-25 00:33:02 UTC

Type the Esc key to interrupt automatic startup
Boot device: /sd/sdhci@d4281000/disk:\boot\olpc.fth  Arguments:
Trying last:\boot\ecfw4.zip
Trying last:\boot\ecfw.zip
Trying last:\boot\bootfw4.zip
  FW found - current FW is up-to-date
Trying last:\boot\runos4.zip
  OS found - Trying last:\boot\runrd4.zip
  RD found - Uncompressing Linux... done, booting the kernel.
[   38.932342] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
[   38.932342]
[   38.941741] [<c0016c3c>] (unwind_backtrace+0x0/0x128) from [<c0456254>] (dump_stack+0x20/0x24)
[   38.950531] [<c0456254>] (dump_stack+0x20/0x24) from [<c0456370>] (panic+0x94/0x1d8)
[   38.958496] [<c0456370>] (panic+0x94/0x1d8) from [<c002acb0>] (do_exit+0x344/0x7a4)
[   38.966339] [<c002acb0>] (do_exit+0x344/0x7a4) from [<c002b368>] (do_group_exit+0x0/0xd0)
[   38.974700] CPU0: stopping
[   38.977478] [<c0016c3c>] (unwind_backtrace+0x0/0x128) from [<c0456254>] (dump_stack+0x20/0x24)
[   38.986236] [<c0456254>] (dump_stack+0x20/0x24) from [<c0014878>] (handle_IPI+0x108/0x178)
[   38.994659] [<c0014878>] (handle_IPI+0x108/0x178) from [<c00085a0>] (gic_handle_irq+0x68/0x70)
[   39.003448] [<c00085a0>] (gic_handle_irq+0x68/0x70) from [<c000e900>] (__irq_svc+0x40/0x70)
[   39.011932] Exception stack(0xc0669f40 to 0xc0669f88)
[   39.017089] 9f40: c06a7d48 a0000093 00000001 60400100 c0668000 c046060c ffffffff c064b3bc
[   39.025421] 9f60: c0f56c00 562f5842 00000000 c0669f9c c0669f48 c0669f88 c0020960 c00102b4
[   39.033752] 9f80: 60000013 ffffffff
[   39.037292] [<c000e900>] (__irq_svc+0x40/0x70) from [<c00102b4>] (cpu_idle+0x94/0xdc)
[   39.045288] [<c00102b4>] (cpu_idle+0x94/0xdc) from [<c044d3c4>] (rest_init+0x7c/0x94)
[   39.053253] [<c044d3c4>] (rest_init+0x7c/0x94) from [<c0624860>] (start_kernel+0x2b8/0x308)
```

USB device issue reported previously was with your mmp3.img.gz file decompressed and placed on a USB HDD.  It still happens.  Open Firmware doesn't detect the device.  Oddly, if I place an Ubuntu 18.04.3 ISO9660 filesystem on the device, Open Firmware does detect it.

Placing the filesystem on a USB Flash device works better; `dir u:\` works fine.

Booting from the filesystem fails;

```log
ok boot u:\                        
Boot device: /usb/disk:\  Arguments: 
Software Interrupt
ok ftrace
execute           Called from align-byte#               at  fda0d2ac
align-byte#       Called from fillbuf                   at  fda0d2ec
fillbuf           Called from shortseek                 at  fda0d4f8
shortseek         Called from fgets                     at  fda0e110
 fdbe65c0
   Catch frame - SP: fdbfbf58   my-self: fd9f6394   handler: fdbfb690 
catch             Called from read                      at  fda65cd8
read              Called from load                      at  fda65d98
execute           Called from $call-self                at  fda2f43c
$call-self        Called from $call-method              at  fda2f51c
        0
   Catch frame - SP: fdbfbf90   my-self: 0   handler: fdbfb68c 
catch             Called from (boot-read)               at  fda3ff8c
 fd9f6394
(boot-read)       Called from boot-read                 at  fda40014
boot-read         Called from $boot-read                at  fda402f0
$boot-read        Called from boot-load                 at  fda40320
boot-load         Called from $boot                     at  fda405c8
$boot             Called from boot                      at  fda405e8
execute           Called from interpret-do-defined      at  fda0a55c
do-defined        Called from $compile                  at  fda0a500
$compile          Called from (interpret                at  fda0a7c4
   Catch frame - SP: fdbfbfac   my-self: 0   handler: fdbfb69c 
catch             Called from (interact)                at  fda0f250
(interact)        Called from interact                  at  fda0f2ac
[...]
```

lkundrak commented:
> ok boot u:\

`boot u:\boot\olpc.fth` please, or just let it autoboot.

(Is `boot u:\` supposed to work too? If so, I can certainly reproduce the same issue no a XO-1.75)

quozl commented:
Okay, thanks.  Autoboot did run olpc.fth.  `boot.xo` did give /bin/bash as init, only on display not serial port.  `boot` gave no output at all after echoing the command line.  Here is log from serial port, which is same as displayed from "Flash an image" onwards;

```log
CForth built 2020-03-24 09:27 from commit d3370770...
releasing
USB devices:
/usb@d4208000/hub@1,0
/usb@d4208000/hub@1,0/scsi@2,0
/usb@d4208000/hub@1,0/scsi@2,0/disk
OLPC 4B1, 1196 MHz, 1 GiB memory, 8 GB internal storage, S/N SHC23800036
OpenFirmware  Q7C07   EC Firmware 0.5.01   2020-03-26 00:04:37 UTC

Type the Esc key to interrupt automatic startup
Boot device: /usb/disk:\boot\olpc.fth  Arguments: 
Flash an image built from lr/mmp3 branch:

  flash last:\ariel.rom
  flash last:\xo175.rom
  flash last:\xo1.rom
  flash last:\xo4.rom

'boot' will proceed booting
  boot-device: last:\boot\zImage
  boot-file:   root=/dev/sda1 console=ttyS2,115200 fbcon=font:TER16x32 vt.color=0xf0 console=tty0 rootwait

'boot.xo' to check compatibility with the OLPC OS kernel

Type 'help' for more information.

ok boot
Boot device: /usb@d4208000/hub@1,0/scsi@2,0/disk@0:\boot\zImage  Arguments: root=/dev/sda1 console=ttyS2,115200 fbcon=font:TER16x32 vt.color=0xf0 console=tty0 rootwait
```

lkundrak commented:
Thank you for the response.

> Okay, thanks. Autoboot did run olpc.fth. boot.xo did give /bin/bash as init, only on display not serial port.

Yes, that is intended. That kernel is too old for systemd that's in the image. Does the keyboard work? Does USB work? I'm wondering if you could mount sysfs, and tar it up and share it, along with dmesg.

> boot gave no output at all after echoing the command line. Here is log from serial port, which is same as displayed from "Flash an image" onwards;

Uh. I believe this could be because XO-4's console serial seems to be at 0xd4017000, not 0xd4018000. A new image with the following change is at 
https://people.freedesktop.org/~lkundrak/mmp3-v2.img.gz

```
-CONFIG_DEBUG_MMP_UART3=y
+CONFIG_DEBUG_MMP_UART2=y
```

(Note that this kernel won't start on XO-1.75 any more; but I guess the early debug output is going to be useful at this point.)

quozl commented:
> Thank you for the response.
> 
> > Okay, thanks. Autoboot did run olpc.fth. boot.xo did give /bin/bash as init, only on display not serial port.
> 
> Yes, that is intended. That kernel is too old for systemd that's in the image. Does the keyboard work? Does USB work? I'm wondering if you could mount sysfs, and tar it up and share it, along with dmesg.

Yes, built-in keyboard works.  USB keyboard works.  Serial works with `exec < /dev/ttyS2 > /dev/ttyS2 2>&1`.

```log
bash-5.0# mount -t sysfs sysfs /sys
bash-5.0# mount -t proc proc /proc
bash-5.0# mount -o remount,rw /
bash-5.0# dmesg > /dmesg
bash-5.0# find /sys > /sys.files
bash-5.0# tar cfv sysfs.tar /sys
tar: removing leading '/' from member names
sys/
sys/fs/
sys/fs/ext4/
sys/fs/ext4/sda1/
sys/fs/ext4/sda1/inode_readahead_blks
tar: short read
bash-5.0# tar cfv /sysfs-bus.tar /sys/bus  
tar: removing leading '/' from member names
sys/bus/
sys/bus/cpu/
sys/bus/cpu/devices/
sys/bus/cpu/devices/cpu0
sys/bus/cpu/devices/cpu1
sys/bus/cpu/drivers/
tar: can't open '/sys/bus/cpu/uevent': Permission denied
tar: can't open '/sys/bus/cpu/drivers_probe': Permission denied
sys/bus/cpu/drivers_autoprobe
tar: short read
bash-5.0# █
```

Do you have a method for tar of sysfs?

dmesg: http://dev.laptop.org/~quozl/y/1jI2wW.txt
sysfs file list: http://dev.laptop.org/~quozl/y/1jI2x1.txt

quozl commented:
> > boot gave no output at all after echoing the command line. Here is log from serial port, which is same as displayed from "Flash an image" onwards;
> 
> Uh. I believe this could be because XO-4's console serial seems to be at 0xd4017000, not 0xd4018000. A new image with the following change is at
> https://people.freedesktop.org/~lkundrak/mmp3-v2.img.gz
> 
> ```
> -CONFIG_DEBUG_MMP_UART3=y
> +CONFIG_DEBUG_MMP_UART2=y
> ```
> 
> (Note that this kernel won't start on XO-1.75 any more; but I guess the early debug output is going to be useful at this point.)

Good progress.  Kernel messages on display and serial console.  Hangs during boot, and the camera LED turns on pretty much at the same time, or slightly after the last kernel message.

Here is serial console; http://dev.laptop.org/~quozl/y/1jI3OY.txt

As far as I can see, same on display.  No response to keyboard or serial port data.

lkundrak committed:
mmp2/clk: increase MPMU reg range

The ACGR is at the offset of 0x1024, beyond the 4k initially assigned to
the MPMU range.

lkundrak commented:
> Hangs during boot, and the camera LED turns on pretty much at the same time, or slightly after the last kernel message.

Looking at clk.fth it indeed seems that turning on the camera block on MMP3 needs some more work to be enabled than it does on MMP2. I'm wondering if you could try disabling the camera node before booting and see if it helps:

```
ok dev /camera@d420a000  " disabled" " status" string-property  dend
```

If it doesn't help, please try booting this image: https://people.freedesktop.org/~lkundrak/mmp3-v3.img.gz

It enables `CONFIG_DEBUG_DRIVER=y` and adds `ignore_loglevel` to the kernel command line; hopefully that will help us identify which driver causes the hang.

Thank you.

quozl commented:
Disabling the camera fixed it, see [serial console log](http://dev.laptop.org/~quozl/z/1jOx48.txt).

quozl commented:
By the way, I'd appreciate knowing how you constructed the Buildroot and filesystem.  That kind of thing could be useful.

lkundrak commented:
> Disabling the camera fixed it,

Woo hoo, that's good news. This means we need a power domain for the CCIC. My branch with clk-of-mmp2 updtes for 5.8 [1] adds power domains for GPUs and audio on MMP2, I'll see if I could throw in one for the camera as well.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/lkundrak/linux-mmp.git/log/?h=lr/mmp2-clk-audio-gpu

> see serial console log.

Some errors I took notice of:

```
[    2.000226] zforce-ts 5-0050: failed to get x-size property
[    2.011117] zforce-ts: probe of 5-0050 failed with error -22
```

Probably just a missing property in the device tree.

```
[    2.850993] sdhci-pxav3 d4281000.sdhci: error -84 requesting status
[    2.862656] mmcblk1: recovery failed!
[    2.924467] sdhci-pxav3 d4281000.sdhci: error -84 requesting status
[    2.936563] mmcblk1: recovery failed!
[    2.946010] blk_update_request: I/O error, dev mmcblk1, sector 0 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    2.967505] Buffer I/O error on dev mmcblk1, logical block 0, async page read
[    2.985890]  mmcblk1: unable to read partition table
```

No idea about this.

```
[   15.750248] mwifiex_sdio mmc3:0001:1: FW download, write iomem (0) failed @ 45760
[   15.769729] mwifiex_sdio mmc3:0001:1: prog_fw failed ret=0xffffffff
[   15.781692] mwifiex_sdio mmc3:0001:1: info: _mwifiex_fw_dpc: unregister device
```

Not sure. Perhaps the root cause for both issues is the same.

> By the way, I'd appreciate knowing how you constructed the Buildroot and filesystem. That kind of thing could be useful.

It's been constructed with Buildroot [1]. It is super useful for building cross toolchains and customized Linux images with very little effort. I've used it so I could rebuild it with a different kernel without manual tinkering and end up with an image of very reasonable size.

[1] https://buildroot.org/

I also find it useful to build Geode-optimized images for XO-1, because no major distribution seems to support it any more. Now if someone packaged Sugar for it! :)

I've submitted the minimal build scripts for XO laptops upstream, with a newest patch set just today [2]. The actual scripts I've used for the mmp2-test* images here are  here [3].

[2] http://lists.busybox.net/pipermail/buildroot/2020-April/thread.html#280273
[3] https://github.com/hackerspace/olpc-xo175-buildroot/commits/lr/mmp3-test3

quozl commented:
Thanks.

lkundrak committed:
mmp3/galcore: split the /gpu node into 2D and 3D

lkundrak committed:
mmp2/clk: rename mmp2-gpu-power-domain# to mmp2-gpu-pd#

This saves some bits!

lkundrak committed:
olpc/mmp2camera: add power domain on MMP3

This tells the driver to enable the power island first.

lkundrak commented:
Some progress, possibly: the power island support is now in linux-next targeting 5.8. This, along with updates to this pull request could possibly resolve the boot hang. I'm wondering if you could test it, as my board doesn't have a camera hooked on.

The image at https://people.freedesktop.org/~lkundrak/mmp3-v6.img.gz (build off [1]) has a today's linux-next kernel and an updated "xo4.rom" image with OFW build from this (lr/mmp3) branch. Please try flashing the firmware and see if it boots. Note that the camera power island will stay on (perhaps the LED will stay on) until [2] is applied.

[1] https://github.com/hackerspace/olpc-xo175-buildroot/commits/lr/mmp3-test6
[2] https://lore.kernel.org/lkml/20200601192124.172650-1-lkundrak@v3.sk/

quozl commented:
Thanks.  I'm away from my lab at the moment.  I should be back there starting in July.

lkundrak committed:
mmp3/hdmi: avoid accidentally writing to some low addresses

It is not clear what is the point the c! here. What's on the stack at
that point is the index into he packet, but a memory address and
executing in just clobbers something in low memory addresses.

Even if it was and address and not an index, writing back what we just
read does not seem to make much sense either -- if I'm reading this
right.

Let's just drop index to conform with the stack diagram. I don't have
equipment picky enough about AVI info frame to test this; ordinary HDMI
and DVI capable computer monitors don't seem to care.

lkundrak committed:
mmp3/hdmi: add 1024x768@60 mode

This is a regular XGI mode, not CEA; a good fallback choice for regular
computer screens.

lkundrak committed:
mmp3/hdmi: support bpp > 16

lkundrak committed:
olpc/lcd: enable HDMI on startup if LCD0 resolution is supported

This does nothing on the XO, because the resolution of the panel on LCD0
is unlike anything mmp3/hdmi knows about.

On the Dell Ariel board, however, there's no internal panel. There's a
DVI-I port driven by LCD0 via an external encoder and a DVI-D port
driven via LCD1's HDMI encoder. We want to bring both up with some safe
resolution, such as 1024x768, and this looks like a reasonable place to
do just that.

lkundrak committed:
mmp3/hdmi: refer to HDMI DDC bus as "i2c6"

Use the alias instead of the node path. This might seem like a rather
useless thing to do, but using an alias will make this easier to reuse
this on Dell Ariel, where a different bus is used for DCC on digital
video port.

lkundrak committed:
mmp3/hdmi: avoid using encode-gpio

It's defined in i2c-gpio; but some boards may just used native TWSI/I2C
controllers without dragging in i2c-gpio at all.

lkundrak commented:
Hi,

I'm wondering if you could give this another look.

I've now up updated this pull request with what I'm currently running on my Dell Ariel board, running Linux 5.10 and generated a new image for testing: https://people.freedesktop.org/~lkundrak/mmp3-v7.img.gz (40M, built off [1]).

[1] https://github.com/hackerspace/olpc-xo175-buildroot/commits/lr/mmp3-test7

There are OFW images built from this branch (as of 594d26b830954) included in the image, ready for flashing.

I expect this to no longer hang while probing camera on XO-4. Note that the debugging knobs are still enabled earlyprintk is configured for XO-4's serial port, so the boot should be more verbose and the image won't boot on XO-1.75.     
 
Thanks
Lubo

quozl commented:
Thanks.  Tested.  [Serial console log](http://dev.laptop.org/~quozl/y/1kzSgk.txt).  It hangs during boot with camera LED on.  Last messages are;

```log
[   20.126103] device: 'd420a000.camera--6-0021': device_add
[   20.169942] i2c 6-0021: Linked as a sync state only consumer to d420a000.camera
[   20.186444] device: 'd4019000.gpio--6-0021': device_add
[   20.196366] sd 0:0:0:0: Attached scsi generic sg0 type 0
[   20.196405] i2c 6-0021: Linked as a sync state only consumer to d4019000.gpio
[   20.206274] device: '0:0:0:0': device_add
[   20.231570] bus: 'i2c': driver_probe_device: matched device 6-0021 with driver ov7670
[   20.231652] PM: Adding info for No Bus:0:0:0:0
[   20.248992] bus: 'i2c': really_probe: probing driver ov7670 with device 6-0021
[   20.275282] ov7670 6-0021: no pinctrl handle
```

Disabling camera does fix the hang.  [Serial console log](http://dev.laptop.org/~quozl/y/1kzSki.txt).

quozl commented:
Per schematic, camera LED is powered from +2.8V_CAM which is switched by EN_CAM_PWR, which comes from PX2128 pin AH26, ND_ALE_SM_ADV.  Multi-function pin configuration registers set this to alternate function 1 in CForth.  GPIO 150 decimal.

quozl commented:
For interest, something is up with eMMC access; [serial console log](http://dev.laptop.org/~quozl/y/1kzT6I.txt).

lkundrak commented:
Thank you. I think I got this (the camera issue) now; please give this one a try as your time permits: https://people.freedesktop.org/~lkundrak/mmp3-v8.img.gz

It is essentially the same as the last image with one sad little patch to kernel on top: https://github.com/hackerspace/olpc-xo175-buildroot/commit/a638f2831664545c95a0e9ea063964fe9108ccb4

> Per schematic, camera LED is powered from +2.8V_CAM which is switched by EN_CAM_PWR, which comes from PX2128 pin AH26, ND_ALE_SM_ADV. Multi-function pin configuration registers set this to alternate function 1 in CForth. GPIO 150 decimal.

That seems to be the same as on XO-1.75, so my guess is that powering the sensor up and down should work just fine on XO-4 as well.

> For interest, something is up with eMMC access; [serial console log](http://dev.laptop.org/~quozl/y/1kzT6I.txt).

Oh no :(

```
$ errno 84
EILSEQ 84 Invalid or incomplete multibyte or wide character
```

Never seen that one. A quick grep around the mmc subsystem suggests that it indicates a CRC error.

I'll look into this later; I have no idea why could this be. My board has an eMMC that generally works fine; that is unless an USB UART adapter is connected. My guess was that this has to do something with these [electrical issues with mmc controller](https://github.com/MitchBradley/cforth/blob/master/src/app/arm-xo-cl4/app.fth#L23), but I didn't look further into that. The symptoms here are different though; the eMMC wouldn't be detected at all.

quozl commented:
Thanks.  Worked fine without camera disabled.  [Serial console log](http://dev.laptop.org/~quozl/y/1l4boS.txt).

lkundrak committed:
olpc/sdhci: add vmmc-supply for eMMC on XO-4

Unlike on XO-1.75, on XO-4, power to eMMC is controlled by a GPIO and the
mainline kernel doesn't follow the power-gpios property. Instead, it
needs an instance of a fixed regulator, defined by
regulator/fixed-regulator.txt.

Create one and bind it to MMC controller eMMC card is connected to
following the regulator/regulator.txt and mmc/mmc.txt bindings.

lkundrak commented:
I've now added 1422c0217cf4ba97eeaa6aaa580bc49345b190b0 that could perhaps make the eMMC work.

Prebuilt image at https://people.freedesktop.org/~lkundrak/q7c07-emmc.rom; plain http mirror here:
```
flash http:\\v6.sk\~lkundrak\q7c07-emmc.rom
```

quozl commented:
Thanks.  That fixed it.  [Serial console log](http://dev.laptop.org/~quozl/y/1l5HBQ.txt).

niclashoyer commented:
@quozl @lkundrak 

sorry to revive this old PR. I'm glad I found this. I've got my hands on an XO-4 and tried your buildroot image. It boots fine from a usb stick, but somehow it does not boot from an sdcard. For me the boot hangs and the mmc1 is never found. As soon as I get a proper serial console I'll also post logs.

```
mmc1: error -84 whilst initialising SDIO card
```

Does the kernel still have modifications? A stock armhf debian kernel (as suggested [here](https://linux-mmp.docs.kernel.org/olpc-xo.html#installing-debian)) does not boot for me. It just hangs after initrd is loaded.

quozl commented:
As far as I know, the kernel changes we made for our product were not upstreamed in full before funding for the work stopped.
